### ISSUE TITLE
Fail to parse the sentence "to be continued"

### ISSUE THREAD

**Issue by leungmanhin** [opencog/opencog#2376]
When I do `(nlp-parse "to be continued")`, it crashes with this error:
```
ERROR: In procedure opencog-extension:
ERROR: Throw to key `C++-EXCEPTION' with args `("cog-fc" "Backtrace:\nIn system/repl/repl.scm:\n 191: 19 [#<procedure 1a43690 at system/repl/repl.scm:181:18 ()>]\nIn ice-9/boot-9.scm:\n 157: 18 [catch #t ...]\nIn system/vm/trap-state.scm:\n 172: 17 [with-default-trap-handler #<procedure debug-trap-handler (frame trap-idx trap-name)> ...]\nIn ice-9/boot-9.scm:\n1645: 16 [%start-stack #t #<procedure 1a7ec30 at system/repl/repl.scm:192:32 ()>]\n1650: 15 [#<procedure 1a7eb40 ()>]\nIn system/repl/repl.scm:\n 148: 14 [with-stack-and-prompt #<procedure 1a6e0c0>]\nIn ice-9/boot-9.scm:\n1645: 13 [%start-stack #t #<procedure 1a78da0 at system/repl/repl.scm:149:33 ()>]\n1650: 12 [#<procedure 1a7eb10 ()>]\nIn opencog/nlp/chatbot/chat-utils.scm:\n 263: 11 [nlp-parse \"to be continued\"]\nIn srfi/srfi-1.scm:\n 578: 10 [map #<procedure interpret (parse-node)> (#)]\nIn opencog/nlp/chatbot/chat-utils.scm:\n  98: 9 [interpret #]\nIn unknown file:\n   ?: 8 [opencog-extension cog-fc (# # #)]\nIn ice-9/boot-9.scm:\n 157: 7 [catch #t #<catch-closure 1700a40> ...]\nIn unknown file:\n   ?: 6 [apply-smob/1 #<catch-closure 1700a40>]\nIn ice-9/boot-9.scm:\n 157: 5 [catch #t #<catch-closure 1700920> ...]\nIn unknown file:\n   ?: 4 [apply-smob/1 #<catch-closure 1700920>]\nIn opencog/nlp/relex2logic/rules/passive.scm:\n  44: 3 [pre-passive-rule # #]\nIn unknown file:\n   ?: 2 [cog-name {()}]\nIn ice-9/boot-9.scm:\n 102: 1 [#<procedure 16ca7c0 at ice-9/boot-9.scm:97:6 (thrown-k . args)> wrong-type-arg ...]\nIn unknown file:\n   ?: 0 [apply-smob/1 #<catch-closure 17008e0> wrong-type-arg ...]\n\nERROR: In procedure apply-smob/1:\nERROR: In procedure cog-name: Wrong type argument in position 1 (expecting opencog atom): ()\nABORT: wrong-type-arg\n (/home/leungmanhin/hansonrobotics/opencog/atomspace/opencog/guile/SchemeEval.cc:1105)")'.
```

**Comment by linas** [comment]
these are typical when something is not an atom i.e. the empty list. Take a look at passive.scm my guess is that word-inst-get-lemma returns nil for one of the words, and so cog-name throws cause it's arg is not an atom.

**Comment by leungmanhin** [comment]
Oh yes, it fails at `(word-inst-get-lemma obj)` because LEFT-WALL is considered as the obj...

**Comment by linas** [comment]
Yeah, that sounds like an Aaron Nitzkin invention.  He altered relex so that whenever there was a phantom of any sort, the left-wall would get used instead.

**Comment by anitzkin** [comment]
Actually, I did that only for one case -- to represent the imperative
subject.   I did, you'll recall, beg you to give relex the capability to
re-write the parse trees instead.

I can't tell you what exactly is going on with "to be continued" since I
don't have the system to play with, but I would strongly suspect that the
root of the problem is that* it's not a sentence*, and none of the
components of the language pipeline are generally equipped to deal with
anything that isn't a grammatically complete sentence.

If you want some wild speculation . . . when you give relex a clause
without a subject, it's going to look to see if it's an imperative and
assign the subject "you" to the left-wall.  Since "to be continued" seems
very likely to get mistakenly parsed as a *passive* sentence, perhaps then
the algorithm turning the subject of a passive a verb into the conceptual
object is getting activated, thereby turning the left-wall into the
object?  That would be my bet.

**Comment by linas** [comment]
It gets parsed as if it were "be continued", and the word "to" is ignored by the parser.

In the long-run, the language-learning project is supposed to deal with the parsing of ungrammatical fragments (twitter being my current favorite example).  Its unclear how language-learning is to be extended to the R2L layer.  I think we'll have to get the base layers working first, and then see how to move up a layer (e.g with pattern-mining on that next layer).

**Comment by linas** [comment]
pull req #2384 should avoid the crash.

**Comment by linas** [comment]
Aaron -- for the sentence, "it is to be continued", if passive is not the right tense, then what should it be?  (for my convenience, it would be great if you could comment at   opencog/link-grammar#355 directly)

**Comment by anitzkin** [comment]
The sentence as a whole is not passive.  There is passive involved in the predicate complement, under one interpretation.  It's complex in an interesting way; I'll see if I can hammer it out for you . . . 

"to be continued" in the sentence is not an object either; it is a predicate complement, which can be seen as equivalent to what you call predicate-adjectives "predadj" in relex now.  You'll recall I introduced the notion of an "adjectival" implicitly in R2L, by taking a variety of grammatical categories, such as prepositional phrases, and treating them as adjectives in R2L, basically, for sentences like:

He prefers the door on the right. ("on the right" is acting as an adjectival modifier of "door")

To break it down, the first relation to classify in "It is to be continued" is that what follows "is" is a "complement" (which I installed in Relex), as in . . . 

It is to be continued.
It is green. (the complement is a predicate adjective)
It is a good show. (the complement is a noun phrase)
It is on next season's schedule. (the complement is a prepositional phrase)

Then "to be continued" is technically a passive infinitive, which you can see by . . . 

It is to continue (an infinitival complement).

Now this is where you begin to get interpretive options.  Either this infinitival passive phrase could be seen as an" adjectival" (which I installed in Relex), as in . . . 

I missed the "to be continued" show. (it's acting like an adjective -- an adjectival)

Or you could see it as an elided sentential complement:

I want the show to be continued.
It is (the show) to be continued.

I think I would go with the first interpretation though.  R2L already knows how to interpret a variety of phrase-types as "adjectivals"--just not this one.  I don't think it occurred to me when I was writing those rules.  

What you need to get straight first is to be able to identify that it's a complement anyway, as in

It is X (whatever X is, it's a complement, being predicated of "It")

Your problem is that this rule competes with your passive rule, I gather.  I would focus on identifying "to be continued" as an infinitival, which happens to be passive, and then recognizing inifnitivals as complements, and then the classification of those complements as predicate-adjectivals I think should belong to R2L.  I installed a fair amount of machinery in Relex and R2L for interpreting complements and adjectivals, but it is definitely incomplete.  Also, you might want to compare these sentence structures with the list of coded and un-coded complement-types that I composed in the document that I wrote about everything I did to Relex and R2L.  I don't even know where that document is now, but I definitely sent a copy to you, Amen, Ben, and Ruiting.

**Comment by linas** [comment]
OK, thanks. link-grammar is parsing some of these correctly; I haven't checked the rest.  Not sure what relex is doing with them. The crash is now fixed; handling this particular sentence is now a low priority.

**Comment by leungmanhin** [comment]
Thanks, the crash is fixed by https://github.com/opencog/opencog/pull/2384

**Comment by linas** [comment]
The "to be continued" is fixed in opencog/link-grammar#378 and will be in version 5.3.10

**Comment by leungmanhin** [comment]
Thanks!
